myhoard/controller.py

```diff
@@ -1310,67 +1409,143 @@ def _process_removed_binlogs(self, binlogs):
             for stream in self.backup_streams:
                 stream.remove_binlogs(binlogs)
 
-    def _purge_old_backups(self):
-        purgeable = [backup for backup in self.state["backups"] if backup["completed_at"] and not backup["recovery_site"]]
-        broken_backups_count = sum(backup["broken_at"] is not None for backup in purgeable)
-        # do not consider broken backups for the count, they will still be purged
-        # but we should only purge when the count of non-broken backups has exceeded the limit.
-        non_broken_backups_count = len(purgeable) - broken_backups_count
+    @staticmethod
+    def count_non_broken_backups(backups: list) -> int:
+        # Every broken full or incremental backup makes the following incremental backups broken
+        sorted_backups = sort_completed_backups(backups)
+        count = len(backups)
+        idx = 0
+        while idx < len(backups):
+            if sorted_backups[idx]["broken_at"] is not None:
+                count -= 1
+                while idx < len(backups) - 1 and sorted_backups[idx + 1]["basebackup_info"].get("incremental"):
+                    count -= 1
+                    idx += 1
+            idx += 1
+        return count
+
+    @enum.unique
+    class NoPurgeReason(str, enum.Enum):
+        no_completed_backups = "no_completed_backups"
+        within_max_amount = "within_max_amount"
+        min_amount_required = "min_amount_required"
+        non_closed_backup = "non_closed_backup"
+        preserved_backup = "preserved_backup"
+        required_backup_within_retention = "required_backup_within_retention"
+        requirements_not_met = "requirements_not_met"
+        active_backup = "active_backup"
+
+    def _get_backup_to_purge(self) -> Backup | NoPurgeReason:
```

Comment:
🙈 Not a fan of that return type.

Reply:
Not ideal I guess, but could not come with with a better idea.